python_scripts/linear_models_ex_03.py

```diff
@@ -14,69 +14,116 @@
-# Appendix - Datasets description section at the end of this MOOC.
-# ```
+# We confirm that all the selected features are numerical.
+#
+# Compute the generalization performance in terms of accuracy of a linear model
+# composed of a `StandardScaler` and a `LogisticRegression`. Use a 10-fold
+# cross-validation with `return_estimator=True` to be able to inspect the
+# trained estimators.
 
 # %%
-import pandas as pd
+# Write your code here.
 
-penguins = pd.read_csv("../datasets/penguins_classification.csv")
-# only keep the Adelie and Chinstrap classes
-penguins = (
-    penguins.set_index("Species").loc[["Adelie", "Chinstrap"]].reset_index()
-)
+# %% [markdown]
+# What is the most important feature seen by the logistic regression?
+#
+# You can use a boxplot to compare the absolute values of the coefficients while
+# also visualizing the variability induced by the cross-validation resampling.
+
+# %%
+# Write your code here.
 
-culmen_columns = ["Culmen Length (mm)", "Culmen Depth (mm)"]
-target_column = "Species"
+# %% [markdown]
+# Let's now work with **both numerical and categorical features**. You can
+# reload the Adult Census dataset with the following snippet:
 
 # %%
-from sklearn.model_selection import train_test_split
+adult_census = pd.read_csv("../datasets/adult-census.csv")
+target = adult_census["class"]
+data = adult_census.drop(columns=["class", "education-num"])
 
-penguins_train, penguins_test = train_test_split(penguins, random_state=0)
+# %% [markdown]
+# Create a predictive model where the categorical data must be one-hot encoded,
+# the numerical data must be scaled, and the predictor is a logistic regression
+# classifier. Use the same 10-fold cross-validation strategy with
+# `return_estimator=True` as above to evaluate this complex pipeline.
+#
+# You may need to increase the number of `max_iter`, which is 100 by default.
 
-data_train = penguins_train[culmen_columns]
-data_test = penguins_test[culmen_columns]
+# %%
+# Write your code here.
+
+# %% [markdown]
+# By comparing the cross-validation test scores of both models fold-to-fold,
+# count the number of times the model using both numerical and categorical
+# features has a better test score than the model using only numerical features.
 
-target_train = penguins_train[target_column]
-target_test = penguins_test[target_column]
+# %%
+# Write your code here.
 
 # %% [markdown]
-# First, let's create our predictive model.
+# For the following questions, you can use the following snippet to get the
+# feature names from the column transformer here named `preprocessor`.
 
 # %%
-from sklearn.pipeline import make_pipeline
-from sklearn.preprocessing import StandardScaler
-from sklearn.linear_model import LogisticRegression
+# preprocessor.fit(data)
+# feature_names = (
+#     preprocessor.named_transformers_["onehotencoder"].get_feature_names_out(
+#         categorical_columns
+#     )
+# ).tolist()
+# feature_names += numerical_columns
+# feature_names
```

Comment:
For info: I had to comment this lines by hand as it was rising a flake8 error `F821 undefined name` when building the exercise from the solution.

We may need to think of a better way to avoid this situation in the future.

Reply:
The issue here is that `preprocessor` is defined in the solution but not in the exercise. So I think it will raise an error when building the jupyterbook.

Reply:
Indeed. We can keep it that way then.

Maybe add a comment to state to reuse the `preprocessor` variable defined in the solution of the previous question.